### PR TITLE
AP-2599: Call HMRC Interface

### PR DESCRIPTION
**Pull request by colinbruce**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-2599)

Add a method of submitting data to the HMRC Interface when a provider confirms their client is employed
This PR changes:-
* **HMRC::CreateResponsesService** - This is triggered from the ApplicantEmployedController and ensures that the application has had the `set_transaction_period` triggered and then creates two;
* **HMRC::Responses** - expressly link to `LegalAidApplications` and add a `url` field for calling the next step, they are then passed to;
* **HMRC::SubmissionWorker** - This Sidekiq job sends the data to the HMRC interface and records the `submission_id` and `url` in the hmrc_response, it then creates a, delayed by 5 seconds;
* **HMRC::ResultWorker** - this calls the hmrc_response.url and retrieves the result from the interface.  If the job is still running it re-tries until it gets a response of `created` or `failed` and then records the resulting JSON in the hmrc_response.response field
* It creates a number of shared classes for use in the new workers to prevent repetition warnings and complexity
* It also adds a section to the admin, application submission page so we can debug the responses from the interface 

## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: spec/services/hmrc/interface/submission_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        it 'raises an exception' do
+          expect {
+            described_class.call(hmrc_response)
+          }.to raise_error HMRC::SubmissionError, /Bad Request: URL:/
+        end
+      end
+
+      context 'connection error' do
+        it 'creates a LegalFramework::Submission error and writes a history record with a backtrace' do
```
Comment (mpw5) [inline spec/services/hmrc/interface/submission_service_spec.rb:65]: I think this is another C&P error 😁

File: spec/services/hmrc/interface/result_service_spec.rb [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+        it 'raises an exception' do
+          expect {
+            described_class.call(hmrc_response)
+          }.to raise_error HMRC::SubmissionError, /Bad Request: URL:/
+        end
+      end
+
+      context 'connection error' do
+        it 'creates a LegalFramework::Submission error and writes a history record with a backtrace' do
```
Comment (mpw5) [inline spec/services/hmrc/interface/result_service_spec.rb:145]: I think this is a C&P error 😁

File: spec/services/hmrc/interface/result_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    end
+
+    context 'when the submission has failed' do
+      let(:expected_status) { 202 } # TODO: Response should be 200, also json response should equal - AP-2639 addresses
+      # {
+      #   data: [
+      #     { correlation_id: hmrc_response.submission_id },
+      #     { error: 'submitted client details could not be found in HMRC service' }
+      #   ]
+      # }
```
Comment (mpw5) [inline spec/services/hmrc/interface/result_service_spec.rb:93]: Can this be removed?

**Review by mpw5 — COMMENTED** [review]
Couple of v minor comments.

File: spec/services/hmrc/interface/result_service_spec.rb (hunk 3 above)
Comment (colinbruce) [reply spec/services/hmrc/interface/result_service_spec.rb:93]: It was kind of the continuation of the TODO ...also json response should equal
For when the updates get made to the interface

File: spec/services/hmrc/interface/result_service_spec.rb (hunk 2 above)
Comment (colinbruce) [reply spec/services/hmrc/interface/result_service_spec.rb:145]: How do you feel about this being named `SubmissionError`? As it became used in submission _and_ result, would `HMRC::InterfaceError` be better?

Comment (colinbruce) [reply spec/services/hmrc/interface/result_service_spec.rb]: I deleted these and refactored the name anyway!

File: spec/services/hmrc/interface/result_service_spec.rb (hunk 3 above)
Comment (mpw5) [reply spec/services/hmrc/interface/result_service_spec.rb:93]: Sorry, didn't spot the comment on the previous line 🤦 That makes sense to me.

**Review by stephenrichards — APPROVED** [review]

**Pull request merged**